### ISSUE TITLE
Project: OpenType math font support

### ISSUE THREAD

**Issue by mgubi** [XmacsLabs/mogan#523]
At the moment TeXmacs (and Mogan) are able to use only very few math fonts (like Stix, TeX gyre) and this support is hardwired in the C++ files and does not take into account all the informations available in the font files.

Some time ago I've started to implement some OpenType font support for TeXmacs, here: https://github.com/mgubi/texmacs/tree/wip-unicode-math

If some other developer is interested we could try to make it working in Mogan. TeXmacs still depends very much on old font formats while there are now available many free fonts in OpenType format which make it easier to support professional math typesetting. In particular there are Latin Modern fonts which can replace our old TeX fonts, and possibly improve speed since there will be less font searching. 

One side goal could be to get rid of all TeX fonts and distribute only the OpenType variants. This is for example what ConTeXt does.:

https://wiki.contextgarden.net/ConTeXt_distribution%27s_Fonts

**Comment by darcy-shen** [comment]
Last time, I tried to upgrade the old TeX fonts, but failed. And discussed with @YiqiXu , he agree that I can set the Stix font as the default math font.

@mgubi Do you think setting Stix font as the default Math font is a good idea for Mogan? Or is there a better choice?

**Comment by mgubi** [comment]
Latin modern is an OTF replacement for the standard TeX fonts. The problem is that TeXmacs does not typeset well yet any OTF math font but only some of them which are hardcoded (Stix, Pagella, ...). I personally do not like Stix. Ideally We need to support uniformly OTF fonts, as I propose here.

**Comment by YiqiXu** [comment]
To clarify, I suggest setting STIX2 as the default fallback font for Mogan. STIX2 supports the largest variety of mathematical symbols compared to other popular unicode-math fonts. The current default font, TeXmacs Computer Modern, has limited support for math symbols. To type unsupported symbols, I have to either set a macro specifying a font or change the document font completely. In both cases, STIX is usually sufficient.

In my experience, this situation often causes confusion or even frustration among new users. Therefore, I think a fallback math font would be beneficial, and it also provides an opportunity to improve Mogan's support for STIX2 since STIX2 has almost replaced STIX in LaTeX.

Naturally, a fallback font is only a temporary solution. @vdhoeven's machinery of semantic math fonts (see his paper [Mathematical Font Art](https://www.texmacs.org/joris/fontart/fontart-abs.html)) is a more elegant solution, but unfortunately, I have never fully understood how it works. (If you @mgubi do, I would appreciate an explanation.) In practice, only a limited number of symbols and fonts support this feature. I suggest expanding @vdhoeven's machinery to support a larger number of OTF fonts in the future.

**Comment by mgubi** [comment]
The problem is that right now we do not have a robust way to locate glyphs in a math font. The current machinery is hard-coded by Joris in the font handling code (in smart_fonts but also other places). My unicode_math branch is an attempt to parse the MATH table in the OTF fonts and use it: it contains the placements within the font of the math glyphs (e.g. the various sizes of parentheses, the spacings, the extensible elements). If this become robust we can use any OTF font for math without hardcoded stuff. We would maybe still like to fine tune some spacings but it would be more robust to add other fonts. I would be happy to discuss with people interested what I learned about this. It would be great if we can improve my branch. In the meanwhile you can give a look at its changes to see the points in the TeXmacs codebase where specific code for maths fonts is handled/needed.

**Comment by mgubi** [comment]
Let me expand my comment: currently TeXmacs is unable to use a font like STIX2 without writing some specialised code. What code: one can look at the source to see how Joris handles fonts like "Pagella" or the "TeX Gyre" family. Many parameters and glyph indexes are hardcoded and they vary from font to font. To support STIX2 as a fallback font one would need to hardcode similar information. But all this informations are already present in the font in its MATH table which currently TeXmacs ignore. My branch contains code to read the table and take the info into account. It was working fine the last time I was working on it, but requires still some more work to be good enough for production. Also maybe some restructuring of the font handling code which is not easy to do in TeXmacs but I hope is easier in Mogan where we can experiment a bit more.

**Comment by darcy-shen** [comment]
>  currently TeXmacs is unable to use a font like STIX2 without writing some specialised code. 

Those specialized code should be maintained in scheme code as a font plugin. That's my plan.

**Comment by mgubi** [comment]
> > currently TeXmacs is unable to use a font like STIX2 without writing some specialised code.
> 
> Those specialized code should be maintained in scheme code as a font plugin. That's my plan.

It is not clear this is possible. Ideally I agree with you but I think most of these code paths are performance critical and would slow down typesetting to put them in scheme, unless one design a reliable caching mechanism.

**Comment by darcy-shen** [comment]
>  but I think most of these code paths are performance critical 

Yes, just improved the performance a lot in 11_17 for Mogan v1.2.1.
